### PR TITLE
Enable Trades, Candles and Functions handling for real time connectio…

### PR DESCRIPTION
**Pull request by filip-prime**
…ns. Require ConnectionId (InstanceId) for each request and set it as part of the rabbitMq queue name upon rabbit subscription

### CODE REVIEW

**Review by iliyan-ivanov-prime — APPROVED** [review]

File: src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Handlers/CandlesWebSocketHandler.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+namespace Lykke.AlgoStore.Api.RealTimeStreaming.DataStreamers.WebSockets.Handlers
+{
+    public class CandlesWebSocketHandler : WebSocketHandlerBase<CandleChartingUpdate>
+    {
+        private readonly RealTimeDataSourceBase<CandleChartingUpdate> _candlesListener;
+
+        public CandlesWebSocketHandler(RealTimeDataSourceBase<CandleChartingUpdate> candlesListener, ILog log) : base(log)
+        {
+            _candlesListener = candlesListener;
+            Messages = candlesListener.Select(t => t);
```
Comment (ghost) [inline src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Handlers/CandlesWebSocketHandler.cs]: What is the point of this select?

File: src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Handlers/FunctionsWebSocketHandler.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+namespace Lykke.AlgoStore.Api.RealTimeStreaming.DataStreamers.WebSockets.Handlers
+{
+    public class FunctionsWebSocketHandler : WebSocketHandlerBase<FunctionChartingUpdate>
+    {
+        private readonly RealTimeDataSourceBase<FunctionChartingUpdate> _functionValuesListener;
+
+        public FunctionsWebSocketHandler(RealTimeDataSourceBase<FunctionChartingUpdate> functionValuesListener, ILog log) : base(log)
+        {
+            _functionValuesListener = functionValuesListener;
+            Messages = functionValuesListener.Select(t => t);
```
Comment (ghost) [inline src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Handlers/FunctionsWebSocketHandler.cs]: Same here

File: src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Handlers/TradesWebSocketHandler.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+using System.Threading.Tasks;
+using Common.Log;
+using Lykke.AlgoStore.Algo.Charting;
+using Lykke.AlgoStore.Api.RealTimeStreaming.Filters;
+using Lykke.AlgoStore.Api.RealTimeStreaming.Sources;
+using Microsoft.AspNetCore.Http;
+
+namespace Lykke.AlgoStore.Api.RealTimeStreaming.DataStreamers.WebSockets.Handlers
+{
+    public class TradesWebSocketHandler : WebSocketHandlerBase<TradeChartingUpdate>
```
Comment (ghost) [inline src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Handlers/TradesWebSocketHandler.cs]: Can these classes be made more generic? The constructors are more or less the same

File: src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Middleware/CandlesWebSocketsMiddleware.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+using Lykke.AlgoStore.Api.RealTimeStreaming.DataStreamers.WebSockets.Handlers;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+
+namespace Lykke.AlgoStore.Api.RealTimeStreaming.DataStreamers.WebSockets.Middleware
+{
+    public class CandlesWebSocketsMiddleware : WebSocketMiddlewareBase
```
Comment (ghost) [inline src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Middleware/CandlesWebSocketsMiddleware.cs]: CandlesWebSocketsMiddleware, FunctionsWebSocketsMiddleware and TradesWebSocketsMiddleware have the exact same copy-pasted code, can you not use only one class?

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Handlers/CandlesWebSocketHandler.cs (hunk 1 above)
Comment (filip-prime) [reply src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Handlers/CandlesWebSocketHandler.cs]: This is kind of the API of Reactive extensions. It actually initializes the observable. I have to do it once, I remember having problems if not doing it. Please see the refactored code.

File: src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Handlers/TradesWebSocketHandler.cs (hunk 3 above)
Comment (filip-prime) [reply src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Handlers/TradesWebSocketHandler.cs]: Please see the overall comment below.

File: src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Middleware/CandlesWebSocketsMiddleware.cs (hunk 4 above)
Comment (filip-prime) [reply src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Middleware/CandlesWebSocketsMiddleware.cs]: Please see the overall comment below.

**Comment by filip-prime** [comment]
I agree with you. The idea of having separate handlers and middlware classes for each of candles, trades and functions was to allow for specific handling for each of them. In fact in earlier version of the code I did have some differences, but after several refactoring I was left with just the constructs of theese handlers. I saw they are the same but decided to leave them as is, as a prompt for implementing specific logic if needed.  Anyway, I have combined them now into a common base generic class. Please the new commit. I've only left Candles being handled individually (it has its own handlers and middleware) because it actually does something differ - candles are not instance specific and there is no need for filter.  Please see the refactored version: https://github.com/LykkeCity/Lykke.AlgoStore/pull/219/commits/c3525455494f8d87a06259960ef1f7bdfd2cb62a

File: src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Middleware/WebSocketMiddleware.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    public class WebSocketMiddleware<T> : WebSocketMiddlewareBase<T>
+    {
+        private readonly RequestDelegate _next;
+
+        public WebSocketMiddleware(RequestDelegate next)
+        {
+            _next = next;
+        }
+
+        public async Task Invoke(HttpContext context, WebSocketHandlerBase<T> webSocketHandler)
```
Comment (ghost) [inline src/Lykke.AlgoStore.Api/RealTimeStreaming/DataStreamers/WebSockets/Middleware/WebSocketMiddleware.cs]: You can allow custom handlers by making the handler a generic constraint inheriting from WebSocketHandlerBase<T>. This way you can avoid having the separate Candles middleware

**Review by ghost — APPROVED** [review]

**Pull request merged**